### PR TITLE
[mono] ILStrip sorts custom attribute table

### PR DESCRIPTION
**Pull request by jandupej**
This prevents custom attribute table corruption by sorting it as the last step when stripping an assembly. Addresses https://github.com/dotnet/runtime/issues/85414.

The PR will have to be backported to net7.0.

### CODE REVIEW

File: src/tasks/MonoTargetsTasks/ILStrip/AssemblyStripper/AssemblyStripper.cs [hunk 1]
```diff
@@ -192,6 +203,15 @@ void PatchResources()
             }
         }
 
+        void SortCustomAttributes()
```
Comment (akoeplinger) [inline src/tasks/MonoTargetsTasks/ILStrip/AssemblyStripper/AssemblyStripper.cs:206]: can you please add a comment about why we need to sort the custom attributes here?

**Review by akoeplinger — APPROVED** [review]

File: src/tasks/MonoTargetsTasks/ILStrip/AssemblyStripper/AssemblyStripper.cs [hunk 2]
```diff
@@ -12,6 +12,16 @@
 
 namespace AssemblyStripper
 {
+    class CustomAttrRowComparer : IComparer
+    {
+        public int Compare(object left, object right)
+        {
+            CustomAttributeRow row_left = (CustomAttributeRow)left;
+            CustomAttributeRow row_right = (CustomAttributeRow)right;
+            return row_left.Parent.RID.CompareTo(row_right.Parent.RID);
```
Comment (lambdageek) [inline src/tasks/MonoTargetsTasks/ILStrip/AssemblyStripper/AssemblyStripper.cs:21]: This is subtly wrong.  the problem is that `RID` on a cecil metadata token masks out the token type.

We actually have to reconstruct the custom attribute coded-index.
something like:

```csharp
   var leftParentCodedIdx = Utilities.CompressMetadataToken(CodedIndex.HasCustomAttribute, row_left.Parent.RID);
   var rightParentCodedIdx = Utilities.CompressMetadataToken(CodedIndex.HasCustomAttribute, row_right.Parent.RID);
   return leftParentCodedIdex.CompareTo(rightParentCodedIndex);
```

Comment (akoeplinger) [reply src/tasks/MonoTargetsTasks/ILStrip/AssemblyStripper/AssemblyStripper.cs:21]: This fixed the issue I was running into on https://github.com/dotnet/runtime/pull/88167

**Pull request merged**